README.txt
Assetylene, a plugin for Movable Type
======================================

This plugin provides a new "Caption" field when inserting an asset into a
post, and the ability to customize the HTML markup produced for publishing MT
assets.

Usage
------

You can create a custom template module, named "Asset Insertion", and that
template will then be used to create the HTML for any asset you embed
into a post.

MT5 only: Custom fields for the appropriate asset object type
(asset/image/audio/video) are now displayed on the "File Options" dialog
window when inserting and uploading a new asset file into a post.

There are a number of template variables made available to this template:

* upload_html: This is the original markup MT constructed for the
inserted asset.
* enclose: 1 or 0, depending on whether the asset is to be wrapped with a
'form' tag.
* include: 1 or 0, depending on whether the image is to be displayed or
simply linked to.
* thumb: 1 or 0, depending on whether the inline image is a thumbnail
of the original.
* align: The value of the alignment option ('none', 'left', 'right',
'center').
* popup: 1 or 0, depending on whether the thumbnail image is linked
to a popup of the full image.
* caption: The value entered for the caption on the insert options dialog.
* label: The 'label' value of the asset being inserted.
* description: The 'description' value of the asset being inserted.
* asset_id: The numeric ID of the asset being inserted.
* form_tag: The opening HTML form tag MT creates to enclose the
asset. If you output this, be sure to add a closing form tag
as well.
* form_class: The CSS class name(s) applied to the form tag.
* form_style: The CSS styling applied to the form tag.
* img_tag: The MT-constructed img tag being inserted (this is unset
if no image is actually inserted and linked to instead).
* img_alt: The "alt" attribute of the inserted image.
* img_height: The numeric height of the inserted image.
* img_width: The numeric width of the inserted image.
* img_src: The url of the inserted image.
* img_style: Any CSS styling applied to the inline image.
* img_class: Any CSS class name(s) applied to the inline image.
* a_tag: The "a" tag produced to link the image (if any).
* a_href: The "href" attribute value of the "a" tag.
* a_onclick: The "onclick" attribute value of the "a" tag.

Also, the asset itself is in context, so any of the MT Asset tags may be
used to generate the asset markup. And the current blog is also in context.

MT5 only: Custom fields template tags for the appropriate asset object type
(asset/image/audio/video) may also be used to generate the asset markup.

An example template can be found in:

    plugins/Assestylene/asset_insertion.mtml

You can use this as a basis for your own custom template.

Installation
-------------

To install this plugin, drop the files included with this plugin into your
Movable Type directory (under the mt-static/plugins/Assetylene,
plugins/Assetylene locations).

License
--------

This plugin has been released under the terms of the GNU Public License,
version 2.0.


CHANGES.txt
# Version 1.04 - October 3, 2012 (davidphillips / 601am)

    * Don't apply Asset Insertion markup if selecting image asset
      for custom field.

# Version 1.03 - September 24, 2012 (davidphillips / 601am)

    * Set CSS class of 'caption' textbox based on MT version.

    * Integrate support for asset custom fields provided by Movable Type 5.

# Version 1.02 - May 24, 2009 (byrnereese)

    * Improved some regexes so they are less greedy
      http://code.sixapart.com/trac/mtplugins/changeset/1534/trunk/Assetylene

# Version 1.01 - May 1, 2009 (byrnereese)

    * Making small change to the size of the caption box so that it does not
      disrupt the asset options dialog so much.
      http://code.sixapart.com/trac/mtplugins/changeset/1524/trunk/Assetylene

    * Fixed bug in which many of the variables (img_src, etc) were not
      properly being set, and added an onfocus javascript hook for the
      caption textarea.
      http://code.sixapart.com/trac/mtplugins/changeset/1525/trunk/Assetylene
  
# Version 1.0 - December 8, 2008 (bradchoate)

    * Initial release; support for MT 4.2 and later.


plugins/Assetylene/lib/Assetylene/Plugin.pm
package Assetylene::Plugin;

use strict;
use MT 4.2;

sub xfm_src {
	my ($cb, $app, $tmpl) = @_;
    my $old = <<'HTML';
            <select name="site_path" id="site_path" onchange="setExtraPath(this)">
                <option value="1">&#60;<__trans phrase="Site Root">&#62;</option>
            <mt:if name="enable_archive_paths">
                <option value="0"<mt:if name="archive_path"> selected="selected"</mt:if>>&#60;<__trans phrase="Archive Root">&#62;</option>
            </mt:if>
            <mt:if name="extra_paths">
                <mt:loop name="extra_paths">
                <option value="<mt:if name="enable_archive_paths">0<mt:else>1</mt:if>" middle_path="<mt:var name="path" escape="html">"<mt:if name="selected"> selected="selected"</mt:if>><mt:var name="label" escape="html"></option>
                </mt:loop>
            </mt:if>
            </select>
            / <input type="text" name="extra_path" id="extra_path" class="extra-path" value="<mt:var name="extra_path" escape="html">" />
            &nbsp;<a href="javascript:void(0);" mt:command="open-folder-selector"><__trans phrase="Choose Folder"></a>
HTML
    $old = quotemeta($old);
    my $new = <<"HTML";
            <input type="hidden" name="site_path" value="1" />
            <__trans phrase="Site Root">&#62;<input type="text" name="extra_path" id="extra_path" class="extra-path" readonly="readonly" value="<mt:var name="extra_path" escape="html">" />
HTML
    $$tmpl =~ s!$old!$new!;

    $old = <<'HTML';
        <mtapp:setting
            id="site_path"
            label_class="top-label"
            label="<__trans phrase="Upload Destination">"
            hint="<$mt:var name="upload_hint"$>"
            show_hint="1">
HTML
    $old = quotemeta($old);
    $new = <<"HTML";
        <mtapp:setting
            id="site_path"
            label_class="top-label"
            label="<__trans phrase="Upload Destination">"
            show_hint="0">
HTML
    $$tmpl =~ s!$old!$new!;




    1;
}

1;

plugins/Assetylene/lib/Assetylene/L10N/ja.pm
package Assetylene::L10N::ja;

use strict;
use base 'Assetylene::L10N::en_us';
use vars qw( %Lexicon );

## The following is the translation table.

%Lexicon = (
	'Provides a new "Caption" field when inserting an asset into a post, and the ability to customize the HTML markup produced for publishing MT assets.'
	   => 'キャプション・フィールドの追加と、アイテムのマークアップをカスタマイズする機能を提供します。',
	'Sample "Asset Insertion" template module.'
	   => '"Asset Insertion"テンプレートモジュールのサンプルです。',
	'This module used by Assetylene plugin.'
	   => 'このモジュールは Assetylene プラグインにより使用されます。',
	'Sample "Asset Insertion" module for Assetylene. This is a templete module. Not a Template Set.'
	   => 'Assetylene プラグインにより使用される"Asset Insertion"モジュールのサンプルです。テンプレートモジュールのみです。テンプレートセットではありません。',
	'AssetInsertion for Assetylene' => 'Assetyleneプラグイン用"Asset Insertion"モジュール',
	'Insert a caption?' => 'キャプションを挿入',
	'Set alt attribute in image?' => '表示画像のalt属性を指定',
	'Asset Insertion' => 'アイテムを挿入する',
	'Use Lightbox Effect' => 'Lightbox効果を使用する',
	'CleanUp Asset Insert' => 'アイテム挿入をクリーンアップする',
	'No CleanUp' => 'クリーンアップしない',
	'Add Class for Wrap Paragraph tag' => 'pタグで囲み、クラスを追加する',
	'Add Class for Image tag' => 'imgタグにクラスを追加する',
	'Alignment Class' => '位置揃えクラス',
	'Lightbox Selector' => 'Lightboxセレクター',
	'RightAlign' => '右揃え',
	'CenterALign' => '中央揃え',
	'LeftAlign' => '左揃え',
	'Use' => '使用する',
	'Remove Blank' => '空白行を削除',
	'Pass' => '行わない',
	'Insertion Pattern' => '挿入パターン',
	'Limit Image Size' => '画像サイズの制限',
	'Limit Thumbnail Width' => 'サムネイルの幅を制限',
	'Limit Thumbnail Height' => 'サムネイルの高さを制限',
	'Limit Image Width' => '画像の幅を制限',
	'Limit Image Height' => '画像の高さを制限',
	'Set alt text for Image.' => '画像にaltテキストを設定する',
);

1;



plugins/Assetylene/lib/Assetylene/CMS.pm
# This program is distributed under the terms of the
# GNU General Public License, version 2.
#
# $Id: CMS.pm 1534 2009-05-24 23:52:58Z breese $

package Assetylene::CMS;

use strict;

my $post_process_upload;
sub init_app {
    # MT 5.0 or greater
    return unless MT->version_number >= 5.0;

    # Override the MT::CMS::Asset:_post_process method to call asset
    # save_filter, pre_save, and post_save callbacks to process custom
    # fields when asset is saved.
    require MT::CMS::Asset;
    no warnings 'redefine';
    unless ($post_process_upload) {
        $post_process_upload = \&MT::CMS::Asset::_process_post_upload;
        *MT::CMS::Asset::_process_post_upload = \&_process_post_upload;
    }
}

sub asset_options_image {
    my ($cb, $app, $param, $tmpl) = @_;

    # Assertions:
    # 'asset_id' template parameter must be present.
    my $asset_id = $param->{asset_id} or return;

    # Asset object must be loadable
    my $asset = MT::Asset->load( $asset_id ) or return;

    # The 'image_alignment' MT template node must be in
    # the template we're working with to add our field above it.
    my $el = $tmpl->getElementById('image_alignment')
        or return;

    my $opt = $tmpl->createElement('app:setting', {
        id => 'image_caption',
        label => MT->translate('Caption'),
        label_class => 'no-header',
        hint => '',
        show_hint => 0,
    });

    require MT::Util;
    # Encode any special characters as HTML entities, since this
    # description is being placed in an HTML textarea:
    my $caption_safe = MT::Util::encode_html( $asset->description );

    # CSS class of HTML textarea, depending on MT version
    my $textarea_class;
    my $mt_version = MT->version_number;
    if ( $mt_version < 5 ) {              # MT 4.x
        $textarea_class = 'full-width';
    } elsif ( $mt_version < 5.1 ) {       # MT 5.0x
        $textarea_class = 'larger-text';
    } else {                              # MT 5.1+
        $textarea_class = 'text full low';
    }

    # Contents of the app:setting tag:
    $opt->innerHTML(<<HTML);
    <input type="checkbox" id="insert_caption" name="insert_caption"
        value="1" />
    <label for="insert_caption">Insert a caption?</label>
    <div class="textarea-wrapper"><textarea name="caption" style="height: 36px;" rows="2" cols=""
        onfocus="getByID('insert_caption').checked=true; return false;"
        class="$textarea_class">$caption_safe</textarea></div>
HTML
    # Insert new field above the 'image_alignment' field:
    $tmpl->insertBefore($opt, $el);
    # Force the tokens of the template to be reprocessed now that
    # we've manipulated it:
    $tmpl->rescan();
}

sub asset_insert {
    my ($cb, $app, $param, $tmpl) = @_;

    my $blog = $app->blog;

    # Do not change markup if selecting asset for custom field
    return 1 if $param->{edit_field} =~ /customfield/;

    # Assertions:
    # Load the user-defined "Asset Insertion" template module.
    # Currently, this template must be named in English. Look both
    # at the blog and system level for this template.
    my $insert_tmpl = $app->model('template')->load({
        name => 'Asset Insertion', type => 'custom',
        blog_id => [ $blog->id, 0 ] });
    return unless $insert_tmpl;

    my $asset = $tmpl->context->stash('asset');

    # Collect all the elements of the MT generated asset markup
    # so they can be manipulated indepdendently by the user-defined
    # template:
    my $html = $param->{upload_html};
    my ($img_tag) = $html =~ /(<img\b[^>]+?>)/s;
    my ($a_tag) = $html =~ /(<a\b[^>]+?>)/s;
    my ($form_tag) = $html =~ /(<form[^>]+?>)/s;

    $param->{enclose} = 1 if $form_tag;
    $param->{include} = 1 if $app->param('include');
    $param->{thumb} = 1 if $app->param('thumb');
    ($param->{align}) = $app->param('align') =~ m/(\w+)/;
    $param->{caption} = $app->param('insert_caption') ? $app->param('caption') : '';
    $param->{popup} = 1 if $app->param('popup');

    $param->{label} = $asset->label;
    $param->{description} = $asset->description;
    $param->{asset_id} = $asset->id;

    $param->{a_tag} = $a_tag;
    ($param->{a_href}) = $a_tag =~ /\bhref="(.+?)"/s;
    ($param->{a_onclick}) = $a_tag =~ /\bonclick="(.+?)"/s;

    $param->{form_tag} = $form_tag;
    ($param->{form_style}) = $form_tag =~ /\bstyle="([^\"]+)"/s;
    ($param->{form_class}) = $form_tag =~ /\bclass="([^\"]+)"/s;

    $param->{img_tag} = $img_tag;
    ($param->{img_height}) = $img_tag =~ /\bheight="(\d+)"/;
    ($param->{img_width}) = $img_tag =~ /\bwidth="(\d+)"/;
    ($param->{img_src}) = $img_tag =~ /\bsrc="([^\"]+)"/s;
    ($param->{img_style}) = $img_tag =~ /\bstyle="([^\"]+)"/s;
    ($param->{img_class}) = $img_tag =~ /\bclass="([^\"]+)"/s;
    ($param->{img_alt}) = $img_tag =~ /\balt="([^\"]+)"/s;

    $insert_tmpl->param( $param );

    my $ctx = $insert_tmpl->context;
    $ctx->stash('blog', $blog);
    $ctx->stash('blog_id', $blog->id);
    $ctx->stash('local_blog_id', $blog->id);
    $ctx->stash('asset', $asset);

    # Process the user-defined template:
    my $new_html = $insert_tmpl->output;

    if (defined($new_html)) {
        # Replace the MT generated asset markup with the user-defined
        # markup:
        $param->{upload_html} = $new_html;
    }
    else {
        # Template build error: die, so this gets logged (we're in a
        # callback, so it won't be surfaced to the user unfortunately)
        die "Error from Asset Insertion module: " . $insert_tmpl->errstr;
    }
}

sub asset_options {
    my ($cb, $app, $param, $tmpl) = @_;

    my $blog = $app->blog;
    my $blog_id = $blog->id;

    # Assertions:
    # Need MT 5.0 or greater
    return unless MT->version_number >= 5.0;

    # Commercial.pack addon (custom fields) must be present
    return unless MT->component('Commercial');

    # 'asset_id' template parameter must be present.
    my $asset_id = $param->{asset_id} or return;

    # Asset object must be loadable
    my $asset = MT::Asset->load( $asset_id ) or return;


    # Insert asset/image/audio/video custom fields


    # The 'tags' MT template node must be in the template
    # we're working with to add custom fields below it.
    my $el = $tmpl->getElementById('tags')
        or return;

    # createElement() does not pass 'attribute_list' parameter
    # (required for multiple instances of the same attribute
    # such as 'regex_replace'), so we have to directly create
    # an MT::Template::Node element.
    my $custom_fields = MT::Template::Node->new(
        tag => 'app:fields',
        attributes => {
            blog_id => $blog_id,
            object_type => $asset->class, 
            object_id => $asset_id,
            regex_replace => 1,
        },
        attribute_list => [
            [ regex_replace => ['/class="text"/g','class="text full"'] ],
            [ regex_replace => ['/class="text high"/g','class="text full low"'] ],
        ],
    );

    # Insert custom fields below the 'tags' field:
    $tmpl->insertAfter($custom_fields, $el);
    # Force the tokens of the template to be reprocessed now that
    # we've manipulated it:
    $tmpl->rescan();


    # Insert jQuery needed for date/time custom fields

    
    # The 'mt:Include' MT template tag including 'dialog/footer.tmpl' must be
    # in the template we're working with to add setvarblock tag above it.
    my $el2 = ${$tmpl->getElementsByName('dialog/footer.tmpl')}[0]
        or return;

    my $setvarblock_jq = $tmpl->createElement('setvarblock', {
        name => 'jq_js_include',
        append => 1,
    });

    # Contents of the setvarblock tag:
    $setvarblock_jq->innerHTML(<<HTML2);
    jQuery('input.text-date').datepicker({
        dateFormat: 'yy-mm-dd',
        dayNamesMin: [<__trans phrase="_LOCALE_CALENDAR_HEADER_">],
        monthNames: ['- 01','- 02','- 03','- 04','- 05','- 06','- 07','- 08','- 09','- 10','- 11','- 12'],
        showMonthAfterYear: true,
        prevText: '&lt;',
        nextText: '&gt;'
    });
HTML2

    # Insert setvarblock above the 'mt:include' tag:
    $tmpl->insertBefore($setvarblock_jq, $el2);
    # Force the tokens of the template to be reprocessed now that
    # we've manipulated it:
    $tmpl->rescan();


    # Add <link> tag to Commercial.pack .css file for custom fields styles


    # The 'mt:Include' MT template tag including 'dialog/header.tmpl' must be
    # in the template we're working with to add setvarblock tag above it.
    my $el3 = ${$tmpl->getElementsByName('dialog/header.tmpl')}[0]
        or return;

    my $setvarblock_css = $tmpl->createElement('setvarblock', {
        name => 'html_head',
        append => 1,
    });

    # Contents of the setvarblock tag:
    $setvarblock_css->innerHTML(<<'HTML3');
<link rel="stylesheet" href="<mt:var name="static_uri">addons/Commercial.pack/styles-customfields.css" type="text/css" media="screen" title="CustomFields Stylesheet" charset="utf-8" />
HTML3

    # Insert setvarblock above the 'mt:include' tag:
    $tmpl->insertBefore($setvarblock_css, $el3);
    # Force the tokens of the template to be reprocessed now that
    # we've manipulated it:
    $tmpl->rescan();
}

sub asset_options_source {
    my ($cb, $app, $src) = @_;

    # Assertions:
    # Need MT 5.0 or greater
    return unless MT->version_number >= 5.0;

    # Commercial.pack addon (custom fields) must be present
    return unless MT->component('Commercial');

    my $magic_token = '<input type="hidden" name="magic_token" value="<mt:var name="magic_token">" />';
    my $asset_type = '  <input type="hidden" name="asset_type" value="<mt:Asset id="$asset_id"><mt:AssetProperty property="class" escape="html"></mt:Asset>" />';
    
    $$src =~ s/($magic_token)/$1\n$asset_type/;
}

sub _process_post_upload {
    my $app   = shift;
    my %param = $app->param_hash;
    my $asset;
    require MT::Asset;
    $param{id} && ( $asset = MT::Asset->load( $param{id} ) )
        or return $app->errtrans("Invalid request.");

    my $mt_version = MT->version_number;
    my ( $need_perm_check, $perm_check );

    if ( $mt_version < 5.1 ) {        # MT 5.0x
        if ( $mt_version >= 5.07 ) {
            $need_perm_check = 1;
        }
    } elsif ( $mt_version < 5.2 ) {   # MT 5.1x
        if ( $mt_version >= 5.13 ) {
            $need_perm_check = 1;
        }
    } else {                          # MT 5.2x
        $need_perm_check = 1;
    }

    if ($need_perm_check) {
        $perm_check =
            ( $app->can('edit_assets') || $asset->created_by == $app->user->id );
    }

    unless ( $need_perm_check && !$perm_check ) {
        my $original = $asset->clone; # For post_save callback
        $asset->label( $param{label} )             if $param{label};
        $asset->description( $param{description} ) if $param{description};
        if ( $param{tags} ) {
            require MT::Tag;
            my $tag_delim = chr( $app->user->entry_prefs->{tag_delim} );
            my @tags = MT::Tag->split( $tag_delim, $param{tags} );
            $asset->set_tags(@tags);
        }

        # Run save_filter callbacks prior to saving asset
        # Custom fields has callback here to validate and
        # pre-process custom fields before saving
        my $filter_result = $app->run_callbacks( 'cms_save_filter.asset', $app );
        return $app->error($app->errstr) unless $filter_result;

        # Added for completeness - no known custom fields callbacks
        $app->run_callbacks( 'cms_pre_save.asset', $app, $asset, $original );

        $asset->save();

        # Run post_save callbacks to save custom fields
        $app->run_callbacks( 'cms_post_save.asset', $app, $asset, $original );
    }

    $asset->on_upload( \%param );
    require MT::CMS::Asset;
    return MT::CMS::Asset::asset_insert_text( $app, \%param );
}

1;
